### ISSUE TITLE
Increasing number of retries and redispaths  that appear after a DDOS attack.

### ISSUE THREAD

**Issue by sklgromek** [haproxy/haproxy#375]
## Output of `haproxy -vv` and `uname -a`

```
Linux proxy-1 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u4 (2019-07-19) x86_64 GNU/Linux

HA-Proxy version 2.0.9-1~bpo9+1 2019/11/17 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.0.9=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 1.1.0k  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.0l  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.22 2016-07-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```

## What's the configuration?

```
global
    log 127.0.0.1 local0
    user haproxy
    group haproxy
    daemon

    nbthread 12

    spread-checks 4
    tune.ssl.default-dh-param 2048

    maxconn 120000
    ulimit-n 500204
    ssl-default-bind-ciphers EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:!MD5;

    ssl-default-bind-options no-sslv3
    tune.ssl.cachesize 600000

    stats socket /tmp/haproxy.sock mode 666


defaults
    mode   http
    log    global
    option httplog
    option http-keep-alive

    retries 3
    option redispatch
    maxconn 120000
    timeout connect 5000
    timeout queue   20000
    timeout client  50000
    timeout server  600000
    timeout http-keep-alive 4s

    option splice-auto

    default-server inter 10s
    default-server fastinter 5s
    default-server downinter 20s
    default-server maxconn 1000

backend klaster
    mode http
    
    balance roundrobin

    cookie SSID prefix

    option forwardfor except 10.202.0.0/24

    option httpchk GET /test.php HTTP/1.1\r\nHost:\ lib.example.com\r\nX-Proto:\ https\r\nX-SSL-Connect:\ Yes\r\nConnection:\ close
    http-check expect status 200


    server app01 10.202.0.201:80 cookie L01 check weight 4
    server app02 10.202.0.202:80 cookie L02 check weight 8
    server app03 10.202.0.203:80 cookie L03 check weight 6
    server app04 10.202.0.204:80 cookie L04 check weight 24
    server app05 10.202.0.205:80 cookie L05 check weight 24
    server app06 10.202.0.206:80 cookie L06 check weight 8
    server app07 10.202.0.207:80 cookie L07 check weight 24
    server app08 10.202.0.208:80 cookie L08 check weight 7
    server app09 10.202.0.209:80 cookie L09 check weight 7
    server app10 10.202.0.210:80 cookie L10 check weight 7
    server app11 10.202.0.211:80 cookie L11 check weight 8
    server app12 10.202.0.212:80 cookie L12 check weight 8
    server app13 10.202.0.213:80 cookie L13 check weight 7
    server app14 10.202.0.214:80 cookie L14 check weight 7
    server app16 10.202.0.216:80 cookie L16 check weight 24  
    server app17 10.202.0.217:80 cookie L17 check weight 24
    server app20 10.202.0.220:80 cookie L20 check weight 24 maxconn 1400
    server app21 10.202.0.221:80 cookie L21 check weight 24 maxconn 1400
    server app26 10.202.0.226:80 cookie L26 check weight 24 maxconn 1400 
    server app27 10.202.0.227:80 cookie L27 check weight 24 maxconn 1400
    server app28 10.202.0.228:80 cookie L28 check weight 24 maxconn 1400
    server app29 10.202.0.229:80 cookie L29 check weight 24 maxconn 1400
    server app30 10.202.0.230:80 cookie L30 check weight 24 maxconn 1400
    server app31 10.202.0.231:80 cookie L31 check weight 24 maxconn 1400
    server app32 10.202.0.232:80 cookie L32 check weight 24 maxconn 1400
    server app33 10.202.0.233:80 cookie L33 check weight 24 maxconn 1400
    server app34 10.202.0.234:80 cookie L34 check weight 24 maxconn 1400
    server app35 10.202.0.235:80 cookie L35 check weight 24 maxconn 1400
    server app36 10.202.0.236:80 cookie L36 check weight 24 maxconn 1400
    server app37 10.202.0.237:80 cookie L37 check weight 24 maxconn 1400
    server app38 10.202.0.238:80 cookie L38 check weight 24 maxconn 1400    
    server app39 10.202.0.239:80 cookie L39 check weight 24 maxconn 1400
    server app40 10.202.0.240:80 cookie L40 check weight 24 maxconn 1400


backend static
    mode http
    balance roundrobin
    
    option httpchk GET /index.html HTTP/1.1\r\nHost:\ lib.example.com
    http-check expect status 200

    server static1 10.202.0.120:80 check weight 5 maxconn 1000
    server static2 10.202.0.121:80 check weight 5 maxconn 1000
    server static3 10.202.0.122:80 check weight 5 maxconn 1000


backend Abuse
    stick-table type ip size 500k expire 1m store gpc0,http_req_rate(5s)

frontend forward_HTTPS
    option forwardfor
    bind *:80    
    bind *:443 ssl crt /etc/haproxy/ssl/cert.pem defer-accept

    http-request add-header X-Proto http if !{ ssl_fc }
    http-request add-header X-SSL-Connect No if !{ ssl_fc }
    http-request add-header X-Proto https if { ssl_fc }
    http-request add-header X-SSL-Connect Yes if { ssl_fc }
 
    acl whitelist_local src  127.0.0.0/8 10.0.0.0/8
    acl whitelist_src_poland src -f /etc/haproxy/pl-aggregated.zone
    acl portal.example.com hdr(host) -i portal.example.com
    acl protected_urls path -i / /f/login // 
    tcp-request content track-sc0 src table Abuse if portal.example.com protected_urls 

    acl acl_abuse sc0_http_req_rate(Abuse) ge 20
    acl acl_flag_abuser sc0_inc_gpc0(Abuse) ge 0

    http-request deny if !whitelist_local !whitelist_src_poland acl_abuse acl_flag_abuser
    http-request deny if { sc0_get_gpc0(Abuse) gt 0 }

    acl not_all_backends nbsrv(klaster) lt 12

    monitor-uri /haproxy_monitor
    monitor fail if not_all_backends

    http-request del-header Proxy

    capture cookie SSID= len 50
    capture request header Host len 64
    capture request header Referer len 64

    acl static_files path_end -i .jpg .png .gif .css .js .webm .pdf
    acl static_folder path_beg -i /images /assets /js /file/css /file/images
    acl portal.example.com hdr(host) -i portal.example.com
    use_backend static if static_files static_folder

    default_backend klaster


```

## Actual behavior

After a larger peak of connections caused by a DDOS attack(even small),
Haproxy has start making retries, and redispatchs to backend, even if the ddos atack is gone.
The problem disappears only after restarting Haproxy service.

At first I thought it had to do with this issue
http://git.haproxy.org/?p=haproxy-2.0.git;a=commitdiff;h=27ebcef
Unfortunately, after the update the problem persisted.

**Comment by wtarreau** [comment]
Hello,

I think I missed that one in the noise. It would be useful to see what logs indicate, if the connections ultimately succeed or not for example. It would be useful to check with netstat (or ss) if you have many idle connections to the servers. It may be possible that you don't have enough source ports to maintain a high number of connections and establish new ones in parallel! And given that you already have ~26000 connections possible to your servers and that the default limit is around 28000 due to ip_local_source_port_ranges, it would make quite some sense to imagine that a number of your incoming requests require a fresh new connection and don't find one.

Also your server timeouts of 10 minutes are absolutely huge and a massive DDoS amplifier, because if the attacker found what request to send there to keep the servers busy, you'll effectively have haproxy sit there and wait for a connection to become available.

In your logs you'll see the time spent in queue, the number of connections in queues, the number of connections per server, backend and frontend. That will help you figure how the issue starts and how it maintains itself over time. I'd strongly suggest to put way shorter timeouts first. If you need to use some long connections for websocket, have a look at "timeout tunnel" instead.

**Comment by sklgromek** [comment]
Hello Willy,
Thanks for your answer.

It is indeed possible that there is a shortage of local ports,
I have set up ip_local_port_range to "2048 65000" value what give me ~62k ports, but during these short DDOS, indeed the number of connections was approaching that limit.

On servers in the backend, there is enough free ports, additionally I have tcp_tw_reuse in kernel enabled.

It is interesting, however, that even after a long time (a few hours after the attack)
still there are resets and redispatch, and only on some servers in the backend.

I have enabled http-keepalive on my apache server, and this problem has not occurred since then.
Thank you for the suggestion, I will adjust the maxconn,  timeouts
